src/proxy/buffer.rs

```diff
@@ -95,72 +147,351 @@ where
+            &mut logging::context_executor(target),
+        )
     }
 }
 
 // === impl MakeFuture ===
 
-impl<F, T, Req, Svc> Future for MakeFuture<F, T, Req>
+impl<F, T, D, Req, Svc> Future for MakeFuture<F, T, D, Req>
 where
     F: Future<Item = Svc>,
     F::Error: Into<Error>,
     Svc: svc::Service<Req> + Send + 'static,
     Svc::Future: Send,
     Svc::Error: Into<Error>,
+    D: Deadline<Req>,
     Req: Send + 'static,
     T: fmt::Display + Send + Sync + 'static,
 {
-    type Item = Buffer<Svc, Req>;
+    type Item = Enqueue<Svc, D, Req>;
     type Error = Error;
 
     fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
         let svc = try_ready!(self.inner.poll().map_err(Into::into));
+        let enq = Enqueue::new(
+            svc,
+            self.deadline.clone(),
+            self.capacity,
+            &mut self.executor,
+        );
+        Ok(enq.into())
+    }
+}
+
+// === impl Enqueue ===
+
+impl<S, D, Req> Enqueue<S, D, Req>
+where
+    S: svc::Service<Req> + Send + 'static,
+    S::Error: Into<Error>,
+    S::Future: Send,
+    Dequeue<S>: svc::Service<Stealer<Req>, Error = Error> + Send,
+    <Dequeue<S> as svc::Service<Stealer<Req>>>::Error: Into<Error>,
+    <Dequeue<S> as svc::Service<Stealer<Req>>>::Future: Send,
+    D: Deadline<Req>,
+    Req: Send + 'static,
+{
+    pub fn new<E>(svc: S, deadline: D, capacity: usize, exec: &mut E) -> Self
+    where
+        E: buffer::WorkerExecutor<Dequeue<S>, Stealer<Req>>,
+    {
+        let inner = buffer::Buffer::with_executor(Dequeue(svc), capacity, exec);
+        Self { deadline, inner }
+    }
+}
+
+impl<S, D, Req> svc::Service<Req> for Enqueue<S, D, Req>
+where
+    Req: Send + 'static,
+    S: svc::Service<Req>,
+    S::Error: Into<Error>,
+    S::Future: Send,
+    D: Deadline<Req>,
+{
+    type Response = S::Response;
+    type Error = Error;
+    type Future = EnqueueFuture<S::Future, Req>;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        self.inner.poll_ready().map_err(Into::into)
+    }
+
+    fn call(&mut self, req: Req) -> Self::Future {
+        let timeout = self.deadline.deadline(&req).map(Delay::new);
+        let holder = Arc::new(Mutex::new(Some(req)));
+        let stealer = Arc::downgrade(&holder);
+
+        EnqueueFuture {
+            holder,
+            timeout,
+            inner: self.inner.call(stealer),
+        }
+    }
+}
+
+impl<S, D, Req> Clone for Enqueue<S, D, Req>
+where
+    S: svc::Service<Req>,
+    S::Error: Into<Error>,
+    D: Clone,
+{
+    fn clone(&self) -> Self {
+        Self {
+            deadline: self.deadline.clone(),
+            inner: self.inner.clone(),
+        }
+    }
+}
+
+// === impl EnqueueFuture ===
+
+impl<Req, F> Future for EnqueueFuture<F, Req>
+where
+    F: Future,
+    F::Error: Into<Error>,
+{
+    type Item = F::Item;
+    type Error = Error;
+
+    fn poll(&mut self) -> Poll<F::Item, Self::Error> {
+        if let Async::Ready(v) = self.inner.poll()? {
```

Comment:
We _do_ only return `Async::NotReady` when the `inner` future has returned `NotReady`

the logic is:
* poll for a response; if it fails, fail. if it succeeds, return it.
* otherwise, if there's no response, and the request still hasn't been dispatched, and if the timeout has expired, return an error.
* otherwise, return not ready. we'll get notified when either the response or timeout future is satisfied.

Reply:
Okay cool thanks for the explanation.